contracts/Items/DEFYForge.sol

```diff
@@ -0,0 +1,241 @@
+    mapping(address => uint256) private operativesPrintJobCount;
+
+    constructor() {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // Forges (mints) a new part and burns the input materials
+    function createForge(
+        IERC1155 itemContract,
+        address operativeAddress,
+        uint256[] calldata inputMaterialsId,
+        uint256[] calldata inputMaterialsAmount,
+        uint256 blueprintId,
+        uint256 outputPartId
+    ) public onlyRole(FORGER_ROLE) returns (bytes32 _printJobId) {
+        // require operative has the blueprint
+        require(
+            itemContract.balanceOf(operativeAddress, blueprintId) != 0,
+            "DEFYForge: Operative does not own blueprint"
+        );
+
+        // require input arrays are not null
+        require(
+            inputMaterialsAmount.length != 0 && inputMaterialsId.length != 0,
+            "DEFYForge: Invalid input materials"
+        );
+
+        // require input materials ids and amount are the same
+        require(
+            inputMaterialsAmount.length == inputMaterialsId.length,
+            "DEFYForge: All arrays must be the same length"
+        );
+
+        // require operative has tokens available
+        for (uint256 i = 0; i < inputMaterialsId.length; i++) {
+            require(
+                itemContract.balanceOf(operativeAddress, inputMaterialsId[i]) >=
+                    inputMaterialsAmount[i],
+                "DEFYForge: Operative does not have suffient materials"
+            );
+        }
+
+        // create print job record
+        bytes32 printJobId = computeNextPrintJobIdForOperative(
+            operativeAddress
+        );
+
+        printJobs[printJobId] = PrintJob(
+            itemContract,
+            operativeAddress,
+            inputMaterialsId,
+            inputMaterialsAmount,
+            blueprintId,
+            outputPartId,
+            false
+        );
+
+        printJobIds.push(printJobId);
+        uint256 currentPrintJobCount = operativesPrintJobCount[
+            operativeAddress
+        ];
+        operativesPrintJobCount[operativeAddress] = currentPrintJobCount + 1;
+
+        // burn input materials
+
+        for (uint256 i = 0; i < inputMaterialsId.length; i++) {
+            itemContract.burnToken(
+                operativeAddress,
+                inputMaterialsId[i],
+                inputMaterialsAmount[i]
+            );
+        }
+
+        // return printJobId
+        return printJobId;
+    }
+
+    // pass in printJobId and mint forged part
+    function completeForge(bytes32 _printJobId)
+        public
+        onlyRole(FORGER_ROLE)
+        returns (address, uint256)
+    {
+        // require printJob has not been completed
+        require(
+            getPrintJob(_printJobId).completed == false,
+            "DEFYForge: Forge has been completed"
+        );
+
+        // get PrintJob struct
+        PrintJob memory _printJob = getPrintJob(_printJobId);
```

Comment:
This should sit above the require above so that you're not getting it twice.  This will use less gas.